lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,40 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
+      end
+
+      # @private
+      class Memoized
+        def initialize
+          @memoized = {}
+          @mutex    = Mutex.new
+        end
+
+        def for(key, &initializer)
+          return @memoized[key][0] if @memoized.key? key
+
+          value = initializer.call
+
+          @mutex.synchronize do
+            memoized = (@memoized[key] ||= [value, Thread.current.object_id])
+            if memoized[1] == Thread.current.object_id
+              # set by call to super in initializer, we override. ie:
+              # let(:title) { super() + ' Child Context' }
+              memoized[0] = value
+            else
+              # some other thread beat us to it, stick with their value
+              memoized[0]
+            end
+          end
+        end
```

Comment:
On the surface, it looks like this implementation uses two different mechanisms to deal with thread safety issues:

* Mutex synchronization.
* `Thread.current_object_id`

Is there a way we can simplify this to one mechanism? FWIW, I played around with this locally and tried this:

``` ruby
def for(key)
  return @memoized[key] if @memoized.key? key

  @mutex.synchronize do
    return @memoized[key] if @memoized.key? key
    @memoized[key] ||= yield
  end
end
```

...which I expected to work, but it didn't, and I realized why: when the logic in one `let` references another `let`, it hits deadlock.  So I realize it may not be simple to get down to just one mechanism :(.

One solution to consider is using [ThreadSafe::Cache#fetch_or_store](https://github.com/ruby-concurrency/thread_safe/blob/v0.3.4/lib/thread_safe/cache.rb#L66-L70), which would "just work", I think, and require no extra synchronization.  That has some other issues, though: we've historically not depended on external gems (with one exception: rspec-expectations uses diff-lcs for diffing) and I'm not sure I want to start doing that.  For example, adding another dependency could prevent users from being able to upgrade RSpec if their bundle already depends on an older version of the external gem than what our version constraint says we support.

More thought is needed, I think...